capa/ghidra/capa_explorer.py

```diff
@@ -0,0 +1,261 @@
+# Integrate capa results with Ghidra UI
+# @author Colton Gabertan (gabertan.colton@gmail.com)
+# @category Python 3.capa
+
+# Copyright (C) 2023 Mandiant, Inc. All Rights Reserved.
+# Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at: [package root]/LICENSE.txt
+# Unless required by applicable law or agreed to in writing, software distributed under the License
+#  is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+import sys
+import json
+import logging
+import pathlib
+from typing import Any, Dict, List
+from contextlib import suppress
+
+from ghidra.app.cmd.label import CreateNamespacesCmd
+from ghidra.program.model.symbol import Namespace, SourceType, SymbolType
+
+import capa
+import capa.main
+import capa.rules
+import capa.render.json
+import capa.ghidra.helpers
+import capa.features.extractors.ghidra.extractor
+
+logger = logging.getLogger("capa_explorer")
+
+
+def add_bookmark(addr, txt, category="CapaExplorer"):
+    """create bookmark at addr"""
+    currentProgram().getBookmarkManager().setBookmark(addr, "Info", category, txt)  # type: ignore [name-defined] # noqa: F821
+
+
+class CapaMatchData:
+    def __init__(
+        self, namespace, scope, capability, locations, node, attack: Dict[Any, Any], mbc: List[Dict[Any, Any]]
+    ):
+        self.namespace = namespace
+        self.scope = scope
+        self.capability = capability
+        self.locations = locations
+        self.node = node
+        self.attack = attack
+        self.mbc = mbc
+
+    def recurse_node(self, node_dict):
+        """pull match descriptions by recursing node dicts"""
+
+        if not node_dict:
+            # mismatched key, skip
+            # ex. {'type':'subscope', 'scope':'basic block'}
+            return ""
+
+        if isinstance(node_dict, int):
+            # Number operands, usually parameters or immediates
+            return hex(node_dict)
+
+        if isinstance(node_dict, str):
+            # expect the "description" key's string value
```

Comment:
Can you expand on what you mean by this comment? It's slightly confusing when considered w/ the statement on line `65`.

Reply:
Please see the updated comments.